File: Java/pattern/composite_pattern/src/com/pattern/LeafComponent_trans.java
package com.pattern;

public class LeafComponent_trans implements IComponent_trans {

	@Override
	public void addChild(IComponent_trans child) {
		System.out.println("I'm leaf, I cannot add child");
	}

	@Override
	public void deleteChild(IComponent_trans child) {
		System.out.println("I'm leaf, I cannot delete child");
	}

	@Override
	public void loopChild() {
		System.out.println("I'm leaf, I cannot loop child");
	}

	@Override
	public void displayClassName() {
		Class<?> enclosingClass = getClass().getEnclosingClass();
		if (enclosingClass != null) {
		  System.out.println(enclosingClass.getName());
		} else {
		  System.out.println(getClass().getName());
		}
	}
}

File: cplusplus/design_pattern/composite_pattern/LeafComponent_trans.cpp
#include "stdafx.h"
#include "LeafComponent_trans.h"


LeafComponent_trans::LeafComponent_trans()
{
}


LeafComponent_trans::~LeafComponent_trans()
{
}

void LeafComponent_trans::addChild(std::shared_ptr<IComponent_trans> child)
{
}

void LeafComponent_trans::deleteChild(std::shared_ptr<IComponent_trans> child)
{
}

void LeafComponent_trans::loopChilds()
{
    displayClassName();
}

void LeafComponent_trans::displayClassName()
{
    std::cout << "LeafComponent_trans::displayClassName()" << std::endl;
}
